torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -240,161 +236,198 @@ BestEffortReplay::BestEffortReplay(
       continue;
     }
 
-    if (t_expr->outputs().size() != r_expr->outputs().size()) {
-      TORCH_INTERNAL_ASSERT(!has_rfactor, err_str);
+    // If there isn't an rfactor id in the replay's inputs and there's a
+    // mismatch in replay_expr's and target_expr's outputs, continue
+    if (target_expr->outputs().size() != replay_expr->outputs().size()) {
+      TORCH_INTERNAL_ASSERT(!replay_has_rfactor_inp, err_str);
       continue;
     }
 
-    if (r_expr->getExprType().value() != t_expr->getExprType().value()) {
-      TORCH_INTERNAL_ASSERT(!has_rfactor, err_str);
+    // If there isn't an rfactor id in the replay's inputs and there's a
+    // mismatch in replay_expr's and target_expr's expression type, continue
+    if (replay_expr->getExprType().value() !=
+        target_expr->getExprType().value()) {
+      TORCH_INTERNAL_ASSERT(!replay_has_rfactor_inp, err_str);
       continue;
     }
 
-    // If the expression is a split, make sure it's split by the same ammount.
-    if (r_expr->getExprType().value() == ExprType::Split) {
-      auto r_split = r_expr->as<Split>();
-      auto t_split = t_expr->as<Split>();
+    // If there isn't an rfactor id in the replay's inputs and there's a
+    // mismatch in replay_expr's and target_expr's split factor (if a split
+    // expr), continue
+    if (replay_expr->getExprType().value() == ExprType::Split) {
+      auto r_split = replay_expr->as<Split>();
+      auto t_split = target_expr->as<Split>();
       if (!r_split->factor()->sameAs(t_split->factor()) ||
           r_split->innerSplit() != t_split->innerSplit()) {
-        TORCH_INTERNAL_ASSERT(!has_rfactor, err_str);
+        TORCH_INTERNAL_ASSERT(!replay_has_rfactor_inp, err_str);
         continue;
       }
     }
 
-    bool missing_input = std::any_of(
-        t_expr->inputs().begin(), t_expr->inputs().end(), [this](Val* inp) {
-          if (inp->getValType() == ValType::IterDomain) {
-            return id_map_.find(inp->as<IterDomain>()) == id_map_.end();
-          }
-          return false;
-        });
+    // Take replay expr inputs out of map:
+    for (size_t t_i = 0; t_i < target_id_inps.size(); t_i++) {
+      auto t_inp = target_id_inps[t_i];
+      auto r_orig_inp = target2replay_id_map_.at(t_inp);
+      auto r_maybe_forwarded_inp = replay_inps[t_i];
 
-    if (missing_input) {
-      TORCH_INTERNAL_ASSERT(!has_rfactor, err_str);
-      continue;
-    }
-    // Take target_domain inputs out of map:
-    for (auto t_inp : ir_utils::filterByType<IterDomain>(t_expr->inputs())) {
-      auto it = id_map_.find(t_inp);
-      if (leaf_ids_.find(it->second) != leaf_ids_.end()) {
-        leaf_ids_.erase(it->second);
+      // Remove original target2replay_it->second if it's in leaf_ids
+      if (leaf_ids_.find(r_orig_inp) != leaf_ids_.end()) {
+        leaf_ids_.erase(r_orig_inp);
+      }
+
+      // Check if we used a forwarded id, if so add forwarded id's to tracking.
+      if (r_orig_inp != r_maybe_forwarded_inp) {
```

Comment:
I don't think this guarantees that a forwarded id is used. In fact, here's an example from `AdvancedIndexing1`:

```
Producer root: iS7{i1} iS8{i3} iS9{i5}
Consumer root: bS10{1} iS11{i1} iS12{i3} iS13{i5}
```

Here, replayCasP is done with the above two tensors.

`is7` is mapped to `iS11`, but it's not located at the same offset in the domain vectors, so `iS11` is added to `forwarded_ids_`, even though no forwarding happens. It seems there's something wrong here, or I may be missing something.

Reply:
So in that case we have:
```
T3[ bS10{1}, iS11{i1}, iS12{i3}, iS13{i5} ] = broadcast( T2[ iS7{i1}, iS8{i3}, iS9{i5} ] )
```
and what's being run is:
```
Replay: T3[ iS30{( ceilDiv(( ceilDiv(( ( ( 1 * i1 ) * i3 ) * i5 ), 128) ), 4) )}, iS31{4}, iS29{128} ] ca_pos( 1 ) produce_pos( 1)
 as T2[ iS36{( ceilDiv(( ceilDiv(( ( i1 * i3 ) * i5 ), 128) ), 4) )}, iS37{4}, iS35{128} ] ca_pos( 1 ) at -1
Forwarded: iS11{i1}
```
What we want is for `T3` to keep the section of the replay it has `( 1 * i1 )` even though `T2` doesn't have that broadcast dimension. So what was forwarded was the root domain `iS11{i1}` in `T3`. It's forwarded to the next transformation so the domain in `T3` with extent `( 1 * i1 )` maps to `iS7{i1}` in `T2`. This way that merge is not "undone" in order for `T3` to match `T2`.

This is the intended behavior.